Error: panic: database is locked

Hello, while runnning the tool like so: `time CloudRecon store -i ipv4_merged.txt -p 443 -t 10 -c 50`. I encountered this error 14 hours later... For context: [ipv4_merged.txt](https://github.com/lord-alfred/ipranges/blob/main/oracle/ipv4_merged.txt)

```panic: database is locked

goroutine 45 [running]:
main.runCloudStore.func1(0x0?)
        /root/go/pkg/mod/github.com/g0ldencybersec/!cloud!recon@v1.0.3/store.go:65 +0x468
created by main.runCloudStore in goroutine 1
        /root/go/pkg/mod/github.com/g0ldencybersec/!cloud!recon@v1.0.3/store.go:49 +0x17e

real    837m40.120s
user    1m50.644s
sys     1m45.807s
```

There is no debug or verbose mode so I can't provide anymore details besides this, I did encounter the same error with `-c 100` which is why I changed it to `-c 50` but to no avail it seems.

Happened again but this one hurts a lot more:

```
panic: database is locked

goroutine 8 [running]:
main.runCloudStore.func1(0x0?)
        /root/go/pkg/mod/github.com/g0ldencybersec/!cloud!recon@v1.0.3/store.go:60 +0x485
created by main.runCloudStore in goroutine 1
        /root/go/pkg/mod/github.com/g0ldencybersec/!cloud!recon@v1.0.3/store.go:49 +0x17e

real    2778m27.455s
user    5m13.007s
sys     5m47.802s
```

--- a/store.go
+++ b/store.go
@@ -23,6 +23,16 @@ type StoreArgs struct {
 	Database    string
 }
 
+// struct to hold data for a database write operation
+type dbWriteRequest struct {
+    ip           string
+    organization string
+    commonName   string
+    san          string
+}
+
+
+
 func runCloudStore(clArgs []string) {
 	args := parseStoreCLI(clArgs)
 
@@ -35,42 +45,53 @@ func runCloudStore(clArgs []string) {
 
 	sqliteDatabase, _ := sql.Open("sqlite3", args.Database) // Open the created SQLite File
 	createTable(sqliteDatabase)                             // Create Database Tables if needed
+	defer sqliteDatabase.Close()                            // Ensure the database is closed when done
 
 	dialer := &net.Dialer{
 		Timeout: time.Duration(args.Timeout) * time.Second,
 	}
 
+	// create a channel for these requests
+	writeRequests := make(chan dbWriteRequest, 100) // Buffered channel
+
 	//Channel for input
 	inputChannel := make(chan string)
 
-	var inputwg sync.WaitGroup
-	for i := 0; i < args.Concurrency; i++ {
-		inputwg.Add(1)
-		go func(db *sql.DB) {
-			for ip := range inputChannel {
-				cert, err := getSSLCert(ip, args.Timeout, dialer)
-				if err != nil {
-					continue
-				} else {
-					names := extractNames(cert)
-					org := cert.Subject.Organization
-					if len(org) > 0 {
-						_, err = db.Exec("INSERT INTO certificates (ip, organization, common_name, san) VALUES (?, ?, ?, ?) ON CONFLICT(ip) DO UPDATE SET organization = excluded.organization, common_name = excluded.common_name, san = excluded.san", ip, org[0], names[0], strings.Join(names[1:], ","))
-						if err != nil {
-							panic(err)
-						}
-					} else {
-						_, err = db.Exec("INSERT INTO certificates (ip, organization, common_name, san) VALUES (?, ?, ?, ?) ON CONFLICT(ip) DO UPDATE SET organization = excluded.organization, common_name = excluded.common_name, san = excluded.san", ip, "NONE", names[0], strings.Join(names[1:], ","))
-						if err != nil {
-							panic(err)
-						}
-					}
-				}
+    // Start the dedicated database writer goroutine
+    go func(db *sql.DB) {
+        for req := range writeRequests {
+            _, err := db.Exec("INSERT INTO certificates (ip, organization, common_name, san) VALUES (?, ?, ?, ?) ON CONFLICT(ip) DO UPDATE SET organization = excluded.organization, common_name = excluded.common_name, san = excluded.san", req.ip, req.organization, req.commonName, req.san)
+            if err != nil {
+                panic(err) // Or handle the error more gracefully
+            }
+        }
+    }(sqliteDatabase)
 
-			}
-			inputwg.Done()
-		}(sqliteDatabase)
-	}
+	var inputwg sync.WaitGroup
+    for i := 0; i < args.Concurrency; i++ {
+        inputwg.Add(1)
+        go func() {
+            defer inputwg.Done()
+            for ip := range inputChannel {
+                cert, err := getSSLCert(ip, args.Timeout, dialer)
+                if err != nil {
+                    continue
+                }
+                names := extractNames(cert)
+                org := "NONE" // Default value if org is not available
+                if len(cert.Subject.Organization) > 0 {
+                    org = cert.Subject.Organization[0]
+                }
+                // Send the write request to the channel
+                writeRequests <- dbWriteRequest{
+                    ip:           ip,
+                    organization: org,
+                    commonName:   names[0], // Assuming names[0] is the common name
+                    san:          strings.Join(names[1:], ","),
+                }
+            }
+        }()
+    }
 
 	intakeFunction(inputChannel, args.Ports, args.Input)
 	close(inputChannel)

